File: clients/ios/AwareChat-iOS/AwareChat-iOS/Core/Persistence/Conversations/ConversationLocalRepository.swift
import Foundation

nonisolated struct LocalConversation: Equatable, Sendable, Identifiable {
  let conversationId: String
  let peer: LocalUser
  let latestMessage: LocalMessage?
  var id: String { conversationId }
}

@MainActor
protocol ConversationLocalRepository: Sendable {
  func conversations() throws -> [LocalConversation]
  func conversation(id: String) throws -> LocalConversation?
  @discardableResult func getOrCreate(with userId: UUID) throws -> LocalConversation
  func observeConversations() -> AsyncThrowingStream<[LocalConversation], any Error>
}

File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/core/persistence/conversations/ConversationLocalRepository.kt
package com.example.awarechat_android.core.persistence.conversations

import com.example.awarechat_android.core.persistence.messages.LocalMessage
import com.example.awarechat_android.core.persistence.users.LocalUser
import java.util.UUID
import kotlinx.coroutines.flow.Flow

data class LocalConversation(
    val conversationId: String,
    val peer: LocalUser,
    val latestMessage: LocalMessage?,
)

interface ConversationLocalRepository {
    suspend fun conversations(): List<LocalConversation>

    suspend fun conversation(id: String): LocalConversation?

    suspend fun getOrCreate(withUserId: UUID): LocalConversation

    fun observeConversations(): Flow<List<LocalConversation>>
}
